core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -576,7 +577,7 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
               val clientResponse = NetworkClientUtils.sendAndReceive(networkClient, request, time)
 
               if (brokerEpochAtRequestTime == kafkaController.brokerEpoch) {
-                val shutdownResponse = clientResponse.responseBody.asInstanceOf[ControlledShutdownResponse]
+                shutdownResponse = clientResponse.responseBody.asInstanceOf[ControlledShutdownResponse]
```

Comment:
This looks good except for one detail. In this if block we set `shutdownResponse` to the value of the actual response from the controller. But if the condition of this if block is false, we could end up calling `notifyControlledShutdownStatus` with the previous response.

Consider this sequence:

1. Broker at brokerEpoch 5 sends ControlledShutdownRequest to controller
2. Broker at brokerEpoch 5 receives ControlledShutdownResponse r1 with an error. Calls notifyControlledShutdownStatus with shutdownResponse=r1.
3. Broker at brokerEpoch 5 sends a second ControlledShutdownRequest to controller
4. Broker loses zookeeper session
5. Broker establishes new zookeeper session with brokerEpoch 16
6. Broker at brokerEpoch 16 receives ControlledShutdownResponse r2 from the controller, but it was sent at epoch 5, so the condition on line 579 fails `brokerEpochAtRequestTime == kafkaController.brokerEpoch => false`
7. Broker calls notifyControlledShutdownStatus with **shutdownResponse=r1**

So I think there needs to be an else block here that sets `shutdownResponse = null`

Reply:
Good catch. Added `shutdownResponse = null` at the top in each iteration.